airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py

```diff
@@ -289,6 +298,62 @@ def properties(self) -> Mapping[str, Any]:
         return props
 
 
+class IncrementalStream(Stream, ABC):
+    """Stream that supports state and incremental read"""
+
+    state_pk = "timestamp"
+    limit = 1000
+
+    @property
+    @abstractmethod
+    def updated_at_fields(self):
+        """Name of the fields associated with the state"""
+
+    @property
+    def state(self) -> Optional[Mapping[str, Any]]:
+        """Current state, if wasn't set return None"""
+        if self._state:
+            return {self.state_pk: str(self._state)}
+        return None
+
+    @state.setter
+    def state(self, value):
+        self._state = pendulum.parse(value[self.state_pk])
+        self._start_date = max(self._state, self._start_date)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._state = None
+
+    def read(self, getter: Callable, params: Mapping[str, Any] = None) -> Iterator:
+        """Apply state filter to set of records, update cursor(state) if necessary in the end"""
+        latest_cursor = None
+        for record in self.read_chunked(getter, params):
```

Comment:
why do we read chunked instead of just paginating over a single request on the entire date range? (i assume it's for a good reason, just not obvious. -- can you leave in a comment?)

Reply:
to track state, there is no guarantee that returned records sorted in ascending order. Having exact boundary we could always ensure we don't miss records between states. In the future, if we would like to save the state more often we can do this every batch